File: src/sophist.cc

#include <node.h>
#include <nan.h>
#include "sophia-cc.h"
#include "sophist.h"
#include "iterator.h"
#include "transaction.h"

namespace sophist {

v8::Persistent<v8::FunctionTemplate> Sophist::constructor;

/**
 * Async open worker.
 */

class OpenWorker : public NanAsyncWorker {
public:
  OpenWorker(
      Sophist *self
    , bool create_if_missing
    , bool read_only
    , uint32_t merge_watermark
    , uint32_t page_size
    , NanCallback *callback
  ) : NanAsyncWorker(callback)
    , self(self)
    , create_if_missing(create_if_missing)
    , read_only(read_only)
    , merge_watermark(merge_watermark)
    , page_size(page_size) {}

  void
  Execute() {
    // TODO: maybe fail?
    if (self->sp->IsOpen()) return;

    sophia::SophiaReturnCode rc = self->sp->Open(
        create_if_missing
      , read_only
      , page_size
      , merge_watermark
    );

    if (sophia::SOPHIA_SUCCESS != rc) {
      errmsg = strdup(self->sp->Error(rc));
    }
  }

private:
  Sophist *self;
  bool create_if_missing;
  bool read_only;
  uint32_t merge_watermark;
  uint32_t page_size;
};

/**
 * Async close worker.
 */

class CloseWorker : public NanAsyncWorker {
public:
  CloseWorker(
      Sophist *self
    , NanCallback *callback
  ) : NanAsyncWorker(callback), self(self) {}

  void
  Execute() {
    if (!self->sp->IsOpen()) {
      errmsg = strdup("Unable to close an unopen database");
      return;
    }

    sophia::SophiaReturnCode rc = self->sp->Close();
    if (sophia::SOPHIA_SUCCESS != rc) {
      errmsg = strdup(self->sp->Error(rc));
    }
  }

private:
  Sophist *self;
};

/**
 * Async get worker.
 */

class GetWorker : public NanAsyncWorker {
public:
  GetWorker(
    Sophist *self
    , char *key
    , NanCallback *callback
  ) : NanAsyncWorker(callback)
    , self(self)
    , key(key) {
    value = NULL;
  }

  void
  Execute() {
    value = self->sp->Get(key);
  }

  void
  HandleOKCallback() {
    v8::Local<v8::Value> err = NanNewLocal<v8::Value>(v8::Null());
    v8::Local<v8::Value> val;
    if (value) {
      val = NanNewLocal<v8::String>(v8::String::New(value));
    } else {
      val = NanNewLocal<v8::Value>(v8::Null());
    }
    v8::Local<v8::Value> argv[2] = { err, val };
    callback->Call(2, argv);
    delete value;
  }

private:
  Sophist *self;
  char *key;
  char *value;
};

/**
 * Async set worker.
 */

class SetWorker : public NanAsyncWorker {
public:
  SetWorker(
      Sophist *self
    , char *key
    , char *value
    , NanCallback *callback
  ) : NanAsyncWorker(callback)
    , self(self)
    , key(key)
    , value(value) {}

  void
  Execute() {
    sophia::SophiaReturnCode rc = self->sp->Set(key, value);
    if (sophia::SOPHIA_SUCCESS != rc) {
      errmsg = strdup(self->sp->Error(rc));
    }
  }

private:
  Sophist *self;
  char *key;
  char *value;
};

/**
 * Async delete worker.
 */

class DeleteWorker : public NanAsyncWorker {
public:
  DeleteWorker(
      Sophist *self
    , char *key
    , NanCallback *callback
  ) : NanAsyncWorker(callback)
    , self(self)
    , key(key) {}

  void
  Execute() {
    sophia::SophiaReturnCode rc = self->sp->Delete(key);
    if (sophia::SOPHIA_SUCCESS != rc) {
      errmsg = strdup(self->sp->Error(rc));
    }
  }

private:
  Sophist *self;
  char *key;
};

/**
 * Async clear worker.
 */

class ClearWorker : public NanAsyncWorker {
public:
  ClearWorker(
      Sophist *self
    , NanCallback *callback
  ) : NanAsyncWorker(callback), self(self) {}

  void
  Execute() {
    sophia::SophiaReturnCode rc = self->sp->Clear();
    if (sophia::SOPHIA_SUCCESS != rc) {
      errmsg = strdup(self->sp->Error(rc));
    }
  }

private:
  Sophist *self;
};

/**
 * Async clear worker.
 */

class CountWorker : public NanAsyncWorker {
public:
  CountWorker(
      Sophist *self
    , NanCallback *callback
  ) : NanAsyncWorker(callback), self(self) {}

  void
  Execute() {
    sophia::SophiaReturnCode rc = self->sp->Count(&count);
    if (sophia::SOPHIA_SUCCESS != rc) {
      errmsg = strdup(self->sp->Error(rc));
    }
  }

  void
  HandleOKCallback() {
    v8::Local<v8::Value> argv[] = {
        NanNewLocal<v8::Value>(v8::Null())
      , NanNewLocal<v8::Number>(v8::Number::New(count))
    };
    callback->Call(2, argv);
  }

private:
  Sophist *self;
  size_t count;
};

Sophist::Sophist(const char *path) : path(path) {
  sp = new sophia::Sophia(path);
}

Sophist::~Sophist() {
  delete path;
}

void
Sophist::Init(v8::Handle<v8::Object> exports) {
  v8::Local<v8::FunctionTemplate> tpl = v8::FunctionTemplate::New(
    Sophist::New
  );
  NanAssignPersistent(v8::FunctionTemplate, constructor, tpl);

  tpl->SetClassName(NanSymbol("Sophist"));
  tpl->InstanceTemplate()->SetInternalFieldCount(1);

  NODE_SET_PROTOTYPE_METHOD(tpl, "open", Open);
  NODE_SET_PROTOTYPE_METHOD(tpl, "close", Close);
  NODE_SET_PROTOTYPE_METHOD(tpl, "get", Get);
  NODE_SET_PROTOTYPE_METHOD(tpl, "getSync", GetSync);
  NODE_SET_PROTOTYPE_METHOD(tpl, "set", Set);
  NODE_SET_PROTOTYPE_METHOD(tpl, "setSync", SetSync);
  NODE_SET_PROTOTYPE_METHOD(tpl, "delete", Delete);
  NODE_SET_PROTOTYPE_METHOD(tpl, "deleteSync", DeleteSync);
  NODE_SET_PROTOTYPE_METHOD(tpl, "count", Count);
  NODE_SET_PROTOTYPE_METHOD(tpl, "clear", Clear);
  NODE_SET_PROTOTYPE_METHOD(tpl, "iterator", NewIterator);
  NODE_SET_PROTOTYPE_METHOD(tpl, "transaction", NewTransaction);

  exports->Set(NanSymbol("Sophist"), tpl->GetFunction());
}

NAN_METHOD(Sophist::New) {
  NanScope();
  size_t pathsize;
  const char *path = NanCString(args[0], &pathsize);
  Sophist *self = new Sophist(path);
  self->Wrap(args.This());
  NanReturnValue(args.This());
}

NAN_METHOD(Sophist::Open) {
  NanScope();
  Sophist *self = node::ObjectWrap::Unwrap<Sophist>(args.This());
  v8::Local<v8::Object> options = args[0].As<v8::Object>();
  v8::Local<v8::Function> fn = args[1].As<v8::Function>();
  bool create_if_missing = NanBooleanOptionValue(
      options
    , NanSymbol("createIfMissing")
    , true
  );
  bool read_only = NanBooleanOptionValue(
      options
    , NanSymbol("readOnly")
    , false
  );
  uint32_t merge_watermark = NanUInt32OptionValue(
      options
    , NanSymbol("mergeWatermark")
    , 100000
  );
  uint32_t page_size = NanUInt32OptionValue(
      options
    , NanSymbol("pageSize")
    , 2048
  );

  OpenWorker *worker = new OpenWorker(
      self
    , create_if_missing
    , read_only
    , merge_watermark
    , page_size
    , new NanCallback(fn)
  );
  NanAsyncQueueWorker(worker);
  NanReturnUndefined();
}

NAN_METHOD(Sophist::Close) {
  NanScope();
  Sophist *self = node::ObjectWrap::Unwrap<Sophist>(args.This());
  v8::Local<v8::Function> fn = args[0].As<v8::Function>();
  CloseWorker *worker = new CloseWorker(self, new NanCallback(fn));
  NanAsyncQueueWorker(worker);
  NanReturnUndefined();
}

NAN_METHOD(Sophist::Get) {
  NanScope();
  Sophist *self = node::ObjectWrap::Unwrap<Sophist>(args.This());
  size_t keysize;
  char *key = NanCString(args[0], &keysize);
  v8::Local<v8::Function> fn = args[1].As<v8::Function>();
  // TODO support `Sophist#get('my key', 1234, function (err, res))`
  GetWorker *worker = new GetWorker(
      self
    , key
    , new NanCallback(fn)
  );
  NanAsyncQueueWorker(worker);
  NanReturnUndefined();
}

NAN_METHOD(Sophist::GetSync) {
  NanScope();
  Sophist *self = node::ObjectWrap::Unwrap<Sophist>(args.This());
  size_t keysize;
  char *key = NanCString(args[0], &keysize);
  char *value = self->sp->Get(key, keysize);
  delete key;
  if (NULL == value) {
    NanReturnValue(v8::Null());
  } else {
    NanReturnValue(v8::String::New(value));
  }
}

NAN_METHOD(Sophist::Set) {
  NanScope();
  Sophist *self = ObjectWrap::Unwrap<Sophist>(args.This());
  size_t keysize;
  size_t valuesize;
  char *key = NanCString(args[0], &keysize);
  char *value = NanCString(args[1], &valuesize);
  v8::Local<v8::Function> fn = args[2].As<v8::Function>();
  SetWorker *worker = new SetWorker(
      self
    , key
    , value
    , new NanCallback(fn)
  );
  NanAsyncQueueWorker(worker);
  NanReturnUndefined();
}

NAN_METHOD(Sophist::SetSync) {
  NanScope();
  Sophist *self = ObjectWrap::Unwrap<Sophist>(args.This());
  size_t keysize;
  size_t valuesize;
  char *key = NanCString(args[0], &keysize);
  char *value = NanCString(args[1], &valuesize);
  sophia::SophiaReturnCode rc = self->sp->Set(
      key
    , value
  );
  if (sophia::SOPHIA_SUCCESS != rc) {
    NanThrowError(self->sp->Error(rc));
  }
  NanReturnUndefined();
}

NAN_METHOD(Sophist::Delete) {
  NanScope();
  Sophist *self = ObjectWrap::Unwrap<Sophist>(args.This());
  size_t keysize;
  char *key = NanCString(args[0], &keysize);
  v8::Local<v8::Function> fn = args[1].As<v8::Function>();
  DeleteWorker *worker = new DeleteWorker(
      self
    , key
    , new NanCallback(fn)
  );
  NanAsyncQueueWorker(worker);
  NanReturnUndefined();
}

NAN_METHOD(Sophist::DeleteSync) {
  NanScope();
  Sophist *self = ObjectWrap::Unwrap<Sophist>(args.This());
  size_t keysize;
  char *key = NanCString(args[0], &keysize);
  sophia::SophiaReturnCode rc = self->sp->Delete(key);
  if (sophia::SOPHIA_SUCCESS != rc) {
    NanThrowError(self->sp->Error(rc));
  }
  NanReturnUndefined();
}

NAN_METHOD(Sophist::Count) {
  NanScope();
  Sophist *self = ObjectWrap::Unwrap<Sophist>(args.This());
  v8::Local<v8::Function> fn = args[0].As<v8::Function>();
  CountWorker *worker = new CountWorker(self, new NanCallback(fn));
  NanAsyncQueueWorker(worker);
  NanReturnUndefined();
}

NAN_METHOD(Sophist::Clear) {
  NanScope();
  Sophist *self = ObjectWrap::Unwrap<Sophist>(args.This());
  v8::Local<v8::Function> fn = args[0].As<v8::Function>();
  ClearWorker *worker = new ClearWorker(self, new NanCallback(fn));
  NanAsyncQueueWorker(worker);
  NanReturnUndefined();
}

NAN_METHOD(Sophist::NewIterator) {
  NanScope();

  v8::Local<v8::Object> options;
  if (args.Length() > 0 && args[0]->IsObject()) {
    options = v8::Local<v8::Object>::Cast(args[0]);
  }

  NanReturnValue(Iterator::NewInstance(
      args.This()
    , options
  ));
}

NAN_METHOD(Sophist::NewTransaction) {
  NanScope();
  NanReturnValue(Transaction::NewInstance(args.This()));
}

} // namespace sophist

File: lib/sophist.js

var fs = require('fs');
var path = require('path');
var DB = require('bindings')('sophist').Sophist;

module.exports = Sophist;

/**
 * Initialize a Sophia binding for `path`.
 *
 * @api public
 * @param {String} path
 */

function Sophist(path) {
  if (!(this instanceof Sophist)) return new Sophist(path);
  this.path = path;
  this.db = new DB(path);
}

/**
 * Open/create the Sophia database with the
 * given `options`, invoking `cb(err)`.
 *
 * Options:
 *
 *  - `createIfMissing` - boolean, default `true`
 *  - `readOnly` - boolean, default `false`
 *  - `gc` - boolean, default `true`
 *  - `pageSize` - number, default `2048`
 *  - `mergeWatermark` - number, default `100000`
 *
 * @api public
 * @param {Object} [options]
 * @param {Function} cb
 */

Sophist.prototype.open = function (options, cb) {
  if ('function' == typeof options) {
    cb = options;
    options = {};
  }
  this.db.open(options, cb);
};

/**
 * Close the Sophia database, invoking `cb(err)`.
 *
 * @api public
 * @param {Function} cb
 */

Sophist.prototype.close = function (cb) {
  this.db.close(cb);
};

/**
 * Set `key=value` in the database, invoking `cb(err)`.
 *
 * @api public
 * @param {String|Buffer|Number} key
 * @param {String|Buffer|Number} value
 * @param {Function} cb
 */

Sophist.prototype.set = function (key, value, cb) {
  key = String(key);
  value = String(value);
  this.db.set(key, value, cb);
};

/**
 * LevelUP compatible `put`.
 *
 * @api public
 * @param {String|Buffer} key
 * @param {String|Buffer} value
 * @param {Object} [options] Unused
 * @param {Function} cb
 */

Sophist.prototype.put = function (key, value, options, cb) {
  cb = 'function' == typeof options
    ? options
    : cb;
  this.set(key, value, cb);
};

/**
 * Synchronously set `key=value`.
 *
 * @api public
 * @param {String|Buffer|Number} key
 * @param {String|Buffer|Number} value
 */

Sophist.prototype.setSync = function (key, value) {
  key = String(key);
  value = String(value);
  this.db.setSync(key, value);
};

/**
 * Get the value of `key`, invoking `cb(err, value)`.
 *
 * @api public
 * @param {String|Buffer|Number} key
 * @param {Object} options Unused
 * @param {Function} cb
 */

Sophist.prototype.get = function (key, options, cb) {
  if (!key) {
    throw new Error('key cannot be `null` or `undefined`');
  }

  key = String(key);
  // levelup support
  cb = 'function' == typeof options
    ? options
    : cb;
  this.db.get(key, cb);
};

/**
 * Synchronously get `key`.
 *
 * @api public
 * @param {String|Buffer|Number} key
 * @return {String}
 */

Sophist.prototype.getSync = function (key) {
  key = String(key);
  return this.db.getSync(key);
};

/**
 * Delete `key` from the database, invoking `cb(err)`.
 *
 * @api public
 * @param {String|Buffer|Number} key
 * @param {Function} cb
 */

Sophist.prototype.delete = function (key, cb) {
  key = String(key);
  this.db.delete(key, cb);
};


/**
 * Synchronously delete `key`.
 *
 * @api public
 * @param {String|Buffer|Number} key
 */

Sophist.prototype.deleteSync = function (key) {
  key = String(key);
  this.db.deleteSync(key);
};

/**
 * LevelUP compatible `del`.
 *
 * @api public
 * @param {String|Buffer} key
 * @param {Object} [options] Unused
 * @param {Function} cb
 */

Sophist.prototype.del = function (key, options, cb) {
  cb = 'function' == typeof options
    ? options
    : cb;
  this.delete(key, cb);
};

/**
 * Get the number of keys in the database, invoking `cb(err, num)`.
 *
 * @api public
 * @param {Function} cb
 */

Sophist.prototype.count = function (cb) {
  this.db.count(cb);
};

/**
 * Clear all keys from the database, invoking `cb(err)`.
 *
 * @api public
 * @param {Function} cb
 */

Sophist.prototype.clear = function (cb) {
  this.db.clear(cb);
};

/**
 * Purge (delete) all incomplete / temporary data
 * from the database.  This operation is **NOT**
 * recoverable.
 *
 * @api public
 * @param {Function} cb
 */

Sophist.prototype.purge = function (cb) {
  var self = this;
  fs.readdir(this.path, function (err, files) {
    if (err) return cb(err);

    var list = files.filter(function (file) {
      return '.log' != file.substr(file.length - 4);
    });

    var pending = list.length;
    var done = false;

    // TODO may cause EMFILEs
    list.forEach(function (file) {
      file = path.join(self.path, file);
      fs.unlink(file, function (err) {
        if (!done && (err || !--pending)) {
          done = true;
          cb(err);
        }
      });
    });
  });
};

/**
 * Create an `Iterator` for the database.
 *
 * You may **not** write data while an iterator is open.
 *
 * Options:
 *
 *  - `reverse` - boolean, default `false`
 *  - `start` - string, default `null`
 *  - `end` - string, default `null`
 *  - `gte` - boolean, default `false`
 *
 * @api public
 * @param {Object} opts
 */

Sophist.prototype.iterator = function (opts) {
  opts = opts || {};

  if (opts.start && 'string' != typeof opts.start) {
    throw new TypeError('Start must be a string');
  }

  if (opts.end && 'string' != typeof opts.end) {
    throw new TypeError('End must be a string');
  }

  return new Iterator(this.db, opts);
};

/**
 * Create a `Transaction` for `db`.
 */

Sophist.prototype.transaction = function () {
  return new Transaction(this.db);
};

/**
 * LevelUP compatible `batch`.
 *
 * @api public
 * @param {Array} operations
 * @param {Object} [options]
 * @param {Function} callback
 */

Sophist.prototype.batch = function (operations, options, callback) {
  if (!arguments.length) return this.transaction();
  var batch = this.transaction(options);

  for (var i = 0, len = operations.length; i < len; i++) {
    var operation = operations[i];
    if ('set' == operation.type || 'put' == operation.type) {
      batch.set(operation.key, operation.value);
    } else {
      batch.delete(operation.key);
    }
  }

  batch.commit(callback);
};

/**
 * Create an `Iterator` for the given `db` with `opts`.
 *
 * @api private
 * @param {SophistBinding} db
 * @param {Object} [opts]
 */

function Iterator(db, opts) {
  this.db = db;
  this.it = this.db.iterator(opts);
}

/**
 * Get the next `key/value` pair from the database,
 * invoking `cb(err, key, value)`.
 *
 * When the iterator has cycled through all data,
 * will call `cb(null, null, null)`.
 *
 * @api public
 * @param {Function} cb
 */

Iterator.prototype.next = function (cb) {
  this.it.next(cb);
};

/**
 * End the iteration cycle, invoking `cb(err)`.
 *
 * @api public
 * @param {Function} cb
 */

Iterator.prototype.end = function (cb) {
  this.it.end(cb);
};

/**
 * Create a `Transaction` for `db`.
 *
 * @api private
 * @param {SophistBinding} db
 */

function Transaction(db) {
  this.db = db;
  this.t = this.db.transaction();
}

/**
 * Commit the transaction operations, invoking `cb(err)`.
 *
 * @api public
 * @param {Function} cb
 */

Transaction.prototype.write =
Transaction.prototype.commit = function (cb) {
  this.t.commit(cb);
};

/**
 * Disregard the transaction operations, invoking `cb(err)`.
 *
 * Committed operations may not be rolled back.
 *
 * @api public
 * @param {Function} cb
 */

Transaction.prototype.rollback = function (cb) {
  // TODO empty operations array?
  this.t.rollback(cb);
};

/**
 * Create a pending `set` operation.  Operation will
 * not be written until `#commit()` is called.
 *
 * @api public
 * @param {String|Buffer|Number} key
 * @param {String|Buffer|Number} value
 * @return {Transaction} for chaining
 */

Transaction.prototype.put =
Transaction.prototype.set = function (key, value) {
  key = String(key);
  value = String(value);
  this.t.set(key, value);
  return this;
};

/**
 * Create a pending `delete` operation.  Operation
 * will not be written until `#commit()` is called.
 *
 * @api public
 * @param {String|Buffer|Number} key
 * @return {Transaction} for chaining
 */

Transaction.prototype.del =
Transaction.prototype.delete = function (key) {
  key = String(key);
  this.t.delete(key);
  return this;
};
